### ISSUE TITLE
Картёжное присвоение зависит от порядка переменных

### ISSUE THREAD

**Issue by SunSerega** [pascalabcnet/pascalabcnet#3171]
(пишу в первую очередь чтобы задокументировать, чтобы в других местах ссылаться)

```
##

procedure tuple_assign<T1,T2>(var a1: T1; var b1: T2; var a2: T1; var b2: T2);
begin
  a1 := a2;
  b1 := b2;
end;

var a := |1|;
var b := |2|;
var c := 3;

var a_ := a;
var b_ := b;

//(a,a[0]) := (b,c); // 3 2
//(a[0],a) := (c,b); // 1 3
//tuple_assign(a,a[0], b,c); // 3 2
//tuple_assign(a[0],a, c,b); // 3 2

//a.Println;
//b.Println;
a_.Println;
b_.Println;
```
В этом коде есть 4 закомментированных варианта присвоения:
```
//(a,a[0]) := (b,c); // 3 2
//(a[0],a) := (c,b); // 1 3
//tuple_assign(a,a[0], b,c); // 3 2
//tuple_assign(a[0],a, c,b); // 3 2
```
Коментарий в конце каждого присвоение - это то, что выводят последние 2 строчки, если раскоментировать именно это присвоение (и никакое другое)

---

Первые 2 присвоения отличаются только порядком переменных - и это меняет результат.
Картёжные присвоения пиаряться как способ присвоить как-бы одновременно:
```
(a,b) := (b,a);
```
Это, по определению, не то же самое что:
```
a := b;
b := a;
```

---

Но как видно из примера выше - текущая реализция не полностью эмулярует эту одновременность.
На практике, используется что то типа:
```
var t1 := b;
var t2 := a;
a := t1;
b := t2;
```
Это работает как ожидалось в простом случае `(a,b) := (b,a)`, но не в более сложном `(a[0],a) := (c,b)` - потому что адрес `a[0]` достаёт только **После** того как выполнилось `a := b`.

---

Но реализовать полную эмуляцию одновременности возможно.
Оба присвоения использующие `tuple_assign` работают одинаково и максимально интуитивно.
Так получается, потому что адрес `a[0]` достаёт чтобы передать `var`-параметром ещё до первого присвоения.

Конечно, генерировать подпрограммы для картёжных присвоений это ужасная идея. Да и исправляет ситуацию не подпрограмма, а её `var`-параметр, аналог `ref`-параметра из C#.
Более конкретным исправлением было бы генерация локальных переменных, аналогичных `ref`-параметрам.
.Net такое позволяет, но в паскале пока нету семантической ноды чтобы эту информацию сохранить, но уже есть все сложные алгоритмы для получения адреса, необходимые чтобы реализовать такую ноду.
И если её таки реализовать - будет сразу возможно добавить это дело и в синтаксис, но это уже долгосрочная идея.

**Comment by miks1965** [comment]
Так может сделать tuple_assign на все случаи жизни? Там всё равно всё инлайнится - и tuple_assign проинлайнится

**Comment by SunSerega** [comment]
Тогда все оптимизации #3162 вылетают в форточку.
Кроме того, C# для такого кода:
```
(a,a[0]) := (b,c);
```
Создаётся `ref`-переменная для `a[0]`, но не для `a`. Если делать через подпрограмму то присвоений будет 3*n.
И если верить презентации из #3162 - эти доп. переменные для присвоений JIT компилятор не убирает, даже если они не нужны, поэтому влияние на производительность тут существенное.

**Comment by miks1965** [comment]
Ну, студент и говорил, что в общем случае в C# может быть 3n присваиваний. У нас - оптимальнее, но этот пример некорректно работает.

Следует отметить, что в Питоне - так же некорректно работает. Там предпочли простую реализацию, не корректную в этом странном случае.

**Comment by SunSerega** [comment]
Нет, в C# только в худшем случае будет 3n присвоений - то же самое и с алгоритмом @nevermind322. Но этот худший случай очень редкий.

В наиболее частом случае в C# будет 2n присвоений, а с алгоритмом @nevermind322 будет в диапазоне 1n..2n присвоений.

**Comment by miks1965** [comment]
@nevermind322 про это и говорил. Но поскольку в его алгоритме решение принимается на синтаксическом уровне, невозможно распознать, что это, массив или нет. Поэтому в случае подобных присваиваний он пессиместически будет ухудшать количество присваиваний до я боюсь 3n и оптимизация рухнет.

**Comment by SunSerega** [comment]
Кхм, ну, если решать это всё на уровне синтаксиса - придётся для всего кроме просто имён переменных (то есть `a[0]` и `a.b`, но не для просто `a`) слева от `:=` - делать `ref`-переменные. И можно `self.a` исключить - потому что `self := ` никогда не работает.

Ну, всё равно, оптимизация @nevermind322 всё равно работает отдельно от создания этих `ref`-переменных, поэтому польза от неё будет даже если всем переменным слева создаст доп. `ref` переменную.

Но опять же, я не понимаю почему это должно решаться на уровне синтаксиса, когда чтобы нормально разобрать - нужна информацию, доступная только во время семантики.
То есть вы ради красивого правила о том что на семантике должно быть минимум логики - связываете себе руки, и провоцируете случаи когда надо идти на уступки и костыли.
На этом этапе может на уровне синтаксиса только объединять присвоение с картежами слева и справа в особую ноду, и разбирать её только во время семантики...

**Comment by miks1965** [comment]
Мы делали подобное в стародавние времена на уровне семантики полностью - генерировали синтаксическое дерево, парсили, обрабатывали ошибки - всё это в семантике. Это был сущий кошмар, провоцирующий другие многочисленные ошибки. Время этого закончилось когда мы ввели именно синтаксические преобразования. Но не всё можно тут сделать.

С введением лямбд генерация синтаксиса на семантике стала вообще невозможно - лямбды трижды сложно обходят синтаксическое дерево.

Чтобы сделать то о чем вы говорите, необходимо полностью менять архитектуру на неизвестную - это катастрофически сложно если вообще возможно нашими ресурсами. Повторюсь, архитектуры подобного нет.

**Comment by SunSerega** [comment]
Ладно, понял проблему.

Но теперь я со свежей головой подумал - и понял что создание `ref`-переменных можно и на уровне синтаксиса исключить в почти всех простых ситуациях.
```
(a.b, c) := (...)
(a[0], c) := (...)
```
Тут в обоих ситуациях не надо создавать `ref` переменные тут не надо. Это потому что `a` всё равно не может поменяться посреди присвоения - самой переменной `a` нету слева от `:=`.

В случае
```
(a.b.c, d.b) := (...)
```
Уже `ref`-переменная нужна для `a.b.c`, потому что `d` может быть той же ссылкой что `a`, а значит `d.b` может быть тем же что `a.b`.

То есть имея даже только имена, без типов, проанализировать и оптимизировать для большинства случаев всё равно реалистично и не так уж сложно.
